app/uk/gov/hmrc/merchandiseinbaggagefrontend/model/declaration/Declaration.scala

```diff
@@ -103,28 +129,69 @@ object JourneyDetails {
-                              maybeAddress: Option[Address] = None,
+                              maybeGoodsDestination: Option[GoodsDestination] = None,
+                              maybeValueWeightOfGoodsExceedsThreshold: Option[Boolean] = None,
+                              goodsEntries: GoodsEntries = GoodsEntries(Seq.empty),
+                              maybeNameOfPersonCarryingTheGoods: Option[Name] = None,
+                              maybeIsACustomsAgent: Option[Boolean] = None,
+                              maybeCustomsAgentName: Option[String] = None,
+                              maybeCustomsAgentAddress: Option[Address] = None,
                               maybeEori: Option[Eori] = None,
                               maybeJourneyDetails: Option[JourneyDetails] = None,
-                              maybeGoodsDestination: Option[GoodsDestination] = None,
-                              maybeValueWeightOfGoodsExceedsThreshold: Option[Boolean] = None) {
-  def toDeclarationIfComplete: Option[Declaration] =
+                              maybeTravellingByVehicle: Option[Boolean] = None,
+                              maybeTravellingBySmallVehicle: Option[Boolean] = None,
+                              maybeRegistrationNumber: Option[String] = None) {
+
+  val maybeCustomsAgent: Option[CustomsAgent] =
+    if (maybeIsACustomsAgent.getOrElse(false)) {
+      for {
+        customsAgentName <- maybeCustomsAgentName
+        customsAgentAddress <- maybeCustomsAgentAddress
+
+        if customsAgentName.trim.nonEmpty
+      } yield CustomsAgent(customsAgentName, customsAgentAddress)
+    } else None
+
+  val journeyDetailsCompleteAndDeclarationRequired: Boolean =
+    maybeJourneyDetails.fold(false){ journeyDetails =>
+      if (journeyDetails.placeOfArrival.requiresVehicleChecks) {
+        if (maybeTravellingByVehicle.contains(false)) {
+          true
+        } else {
+          if (maybeTravellingByVehicle.isEmpty) false
+          else if (!maybeTravellingBySmallVehicle.getOrElse(false)) false
+          else maybeRegistrationNumber.fold(false)(_ => true)
+        }
+      } else true
+    }
+
+  val declarationIfRequiredAndComplete: Option[Declaration] =
     for {
-      name <- maybeName
-      address <- maybeAddress
+      nameOfPersonCarryingTheGoods <- maybeNameOfPersonCarryingTheGoods
+      iAmACustomsAgent <- maybeIsACustomsAgent
       eori <- maybeEori
       journeyDetails <- maybeJourneyDetails
-    } yield
+      goods <- goodsEntries.declarationGoodsIfComplete
+      _ <- maybeGoodsDestination
+
+      if maybeExciseOrRestrictedGoods.contains(false)
+      if maybeValueWeightOfGoodsExceedsThreshold.contains(false)
+      if (iAmACustomsAgent && maybeCustomsAgent.isDefined) || (!iAmACustomsAgent)
+      if journeyDetailsCompleteAndDeclarationRequired
+      if nameOfPersonCarryingTheGoods.firstName.trim.nonEmpty
+      if nameOfPersonCarryingTheGoods.lastName.trim.nonEmpty
+    } yield {
       Declaration(
```

Comment:
few if statements in there :-). Would something like "exists" help? Or even better an Applicative

Reply:
I had meant to remove some of those out as they shouldn't be handled this way. There's an update so now its just:

      if maybeExciseOrRestrictedGoods.contains(false)
      if maybeValueWeightOfGoodsExceedsThreshold.contains(false)
      if maybeCustomsAgent.isDefined || !iAmACustomsAgent
      if journeyDetailsCompleteAndDeclarationRequired

how's that look?